pkg/events/events.go

```diff
@@ -0,0 +1,50 @@
+// Copyright (c) Mainflux
+// SPDX-License-Identifier: Apache-2.0
+
+package events
+
+import (
+	"context"
+	"time"
+)
+
+const (
+	UnpublishedEventsCheckInterval        = 1 * time.Minute
+	ConnCheckInterval                     = 100 * time.Millisecond
+	MaxUnpublishedEvents           uint64 = 1e6
+	MaxEventStreamLen              int64  = 1e9
+)
+
+// Event represents an event.
+type Event interface {
+	// Encode encodes event to map.
+	Encode() (map[string]interface{}, error)
+}
+
+// Publisher specifies events publishing API.
+type Publisher interface {
+	// Publish publishes event to stream.
+	Publish(ctx context.Context, event Event) error
+
+	// StartPublishingRoutine starts a routine that checks for unpublished events
+	// and publishes them to stream.
+	StartPublishingRoutine(ctx context.Context)
```

Comment:
this doesn't need to be part of the interface or exported

Reply:
It is exported so that the different `es` should implement it. As we want the different es for example `Redis streams` or `NATS JS` to implement it as a functionality.

Reply:
do both of them require such a routine with the same signature and some possible es we might use in the future

Reply:
RabbitMQ and Redis don't support using buffers. Hence the implementation for Nats is ignore https://github.com/rodneyosodo/mainflux/blob/d140b10a4936bee0afbadd0c34486e70f5f4f153/pkg/events/nats/publisher.go#L77

Reply:
The function only takes in context for cancellation purposes as the implementation might vary.

Reply:
Since it's purely related to the implementation, let's remove this method from the Publisher interface for the time being.